File: platforms/android/stato/src/main/java/com/facebook/stato/core/StatoArray.java
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
package com.facebook.stato.core;

import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class StatoArray {
  final JSONArray mJson;

  public StatoArray(JSONArray json) {
    mJson = (json != null ? json : new JSONArray());
  }

  public StatoArray(String json) {
    try {
      mJson = new JSONArray(json);
    } catch (JSONException e) {
      throw new RuntimeException(e);
    }
  }

  public StatoDynamic getDynamic(int index) {
    return new StatoDynamic(mJson.opt(index));
  }

  public String getString(int index) {
    return mJson.optString(index);
  }

  public int getInt(int index) {
    return mJson.optInt(index);
  }

  public long getLong(int index) {
    return mJson.optLong(index);
  }

  public float getFloat(int index) {
    return (float) mJson.optDouble(index);
  }

  public double getDouble(int index) {
    return mJson.optDouble(index);
  }

  public boolean getBoolean(int index) {
    return mJson.optBoolean(index);
  }

  public StatoObject getObject(int index) {
    final Object o = mJson.opt(index);
    return new StatoObject((JSONObject) o);
  }

  public StatoArray getArray(int index) {
    final Object o = mJson.opt(index);
    return new StatoArray((JSONArray) o);
  }

  public int length() {
    return mJson.length();
  }

  public List<String> toStringList() {
    final int length = length();
    final List<String> list = new ArrayList<>(length);
    for (int i = 0; i < length; i++) {
      list.add(getString(i));
    }
    return list;
  }

  public String toJsonString() {
    return toString();
  }

  @Override
  public String toString() {
    return mJson.toString();
  }

  @Override
  public boolean equals(Object o) {
    return mJson.toString().equals(o.toString());
  }

  @Override
  public int hashCode() {
    return mJson.hashCode();
  }

  public static class Builder {
    private final JSONArray mJson;

    public Builder() {
      mJson = new JSONArray();
    }

    public Builder put(String s) {
      mJson.put(s);
      return this;
    }

    public Builder put(Integer i) {
      mJson.put(i);
      return this;
    }

    public Builder put(Long l) {
      mJson.put(l);
      return this;
    }

    public Builder put(Float f) {
      mJson.put(Float.isNaN(f) ? null : f);
      return this;
    }

    public Builder put(Double d) {
      mJson.put(Double.isNaN(d) ? null : d);
      return this;
    }

    public Builder put(Boolean b) {
      mJson.put(b);
      return this;
    }

    public Builder put(StatoValue v) {
      return put(v.toStatoObject());
    }

    public Builder put(StatoArray a) {
      mJson.put(a == null ? null : a.mJson);
      return this;
    }

    public Builder put(StatoArray.Builder b) {
      return put(b.build());
    }

    public Builder put(StatoObject o) {
      mJson.put(o == null ? null : o.mJson);
      return this;
    }

    public Builder put(StatoObject.Builder b) {
      return put(b.build());
    }

    public StatoArray build() {
      return new StatoArray(mJson);
    }
  }
}

File: platforms/android/android-stato/src/main/kotlin/org/stato/core/StatoArray.kt
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
package org.stato.core

import org.json.JSONArray
import org.json.JSONObject

class StatoArray(json: JSONArray?) {

  internal val json = json ?: JSONArray()


  constructor(json: String) : this(JSONArray(json))

  fun getDynamic(index: Int): StatoDynamic {
    return StatoDynamic(json.opt(index))
  }

  fun getString(index: Int): String {
    return json.optString(index)
  }

  fun getInt(index: Int): Int {
    return json.optInt(index)
  }

  fun getLong(index: Int): Long {
    return json.optLong(index)
  }

  fun getFloat(index: Int): Float {
    return json.optDouble(index).toFloat()
  }

  fun getDouble(index: Int): Double {
    return json.optDouble(index)
  }

  fun getBoolean(index: Int): Boolean {
    return json.optBoolean(index)
  }

  fun getObject(index: Int): StatoObject {
    val o = json.opt(index)
    return StatoObject(o as JSONObject)
  }

  fun getArray(index: Int): StatoArray {
    val o = json.opt(index)
    return StatoArray(o as JSONArray)
  }

  fun length(): Int {
    return json.length()
  }

  fun toStringList(): List<String> {
    return 0.until(length()).map { getString(it)}
  }

  fun toJsonString(): String {
    return toString()
  }


  override fun toString(): String {
    return json.toString()
  }

  override fun equals(other: Any?): Boolean {
    return json.toString() == other?.toString()
  }


  override fun hashCode(): Int {
    return json.hashCode()
  }

  class Builder {
    private val json: JSONArray = JSONArray()

    fun put(s: String): Builder {
      json.put(s)
      return this
    }

    fun put(i: Int): Builder {
      json.put(i)
      return this
    }

    fun put(l: Long): Builder {
      json.put(l)
      return this
    }

    fun put(f: Float): Builder {
      json.put(if (Float.NaN == f) null else f)
      return this
    }

    fun put(d: Double): Builder {
      json.put(if (Double.NaN == d) null else d)
      return this
    }

    fun put(b: Boolean): Builder {
      json.put(b)
      return this
    }

    fun put(v: StatoValue): Builder {
      return put(v.toStatoObject())
    }

    fun put(a: StatoArray?): Builder {
      json.put(a?.json)
      return this
    }

    fun put(b: Builder): Builder {
      return put(b.build())
    }

    fun put(o: StatoObject?): Builder {
      json.put(o?.json)
      return this
    }

    fun put(b: StatoObject.Builder): Builder {
      return put(b.build())
    }

    fun build(): StatoArray {
      return StatoArray(json)
    }
  }
}
